# DSC 550 - Data Mining
# Week 1 Exercise: Exploring a Pandas Data Frame

In [1]:
#import numpy and pandas libraries
import numpy as np
import pandas as pd

In [25]:
#1) Load the data as a Pandas data frame
df_games = pd.read_csv('video_game_data.csv')

In [3]:
#2) Display the first ten rows of data
df_games.head(10)

Name Platform  Year_of_Release         Genre  \
0                 Wii Sports      Wii           2006.0        Sports   
1          Super Mario Bros.      NES           1985.0      Platform   
2             Mario Kart Wii      Wii           2008.0        Racing   
3          Wii Sports Resort      Wii           2009.0        Sports   
4   Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
5                     Tetris       GB           1989.0        Puzzle   
6      New Super Mario Bros.       DS           2006.0      Platform   
7                   Wii Play      Wii           2006.0          Misc   
8  New Super Mario Bros. Wii      Wii           2009.0      Platform   
9                  Duck Hunt      NES           1984.0       Shooter   

  Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  \
0  Nintendo     41.36     28.96      3.77         8.45         82.53   
1  Nintendo     29.08      3.58      6.81         0.77         40.24   
2  Nintendo     15.68     12.76      3.79         3.29         35.52   
3  Nintendo     15.61     10.93      3.28         2.95         32.77   
4  Nintendo     11.27      8.89     10.22         1.00         31.37   
5  Nintendo     23.20      2.26      4.22         0.58         30.26   
6  Nintendo     11.28      9.14      6.50         2.88         29.80   
7  Nintendo     13.96      9.18      2.93         2.84         28.92   
8  Nintendo     14.44      6.94      4.70         2.24         28.32   
9  Nintendo     26.93      0.63      0.28         0.47         28.31   

   Critic_Score  Critic_Count User_Score  User_Count Developer Rating  
0          76.0          51.0          8       322.0  Nintendo      E  
1           NaN           NaN        NaN         NaN       NaN    NaN  
2          82.0          73.0        8.3       709.0  Nintendo      E  
3          80.0          73.0          8       192.0  Nintendo      E  
4           NaN           NaN        NaN         NaN       NaN    NaN  
5           NaN           NaN        NaN         NaN       NaN    NaN  
6          89.0          65.0        8.5       431.0  Nintendo      E  
7          58.0          41.0        6.6       129.0  Nintendo      E  
8          87.0          80.0        8.4       594.0  Nintendo      E  
9           NaN           NaN        NaN         NaN       NaN    NaN

In [4]:
#3) Find the dimensions of the data frame
df_games.shape

(16719, 16)

There are $16,719$ rows, each corresponding to a different video game. There are $16$ columns, each correspoinding to a feature of the video game in that row.

In [5]:
#4) Top five games by critic score
df_games[['Name','Platform', 'Critic_Score']].sort_values(by = 'Critic_Score', ascending = False).head(5)

Name Platform  Critic_Score
227   Tony Hawk's Pro Skater 2       PS          98.0
57         Grand Theft Auto IV      PS3          98.0
51         Grand Theft Auto IV     X360          98.0
5350               SoulCalibur       DC          98.0
165         Grand Theft Auto V     XOne          97.0

In [6]:
#5) Number of games in each genre
df_games['Genre'].value_counts()

Action          3370
Sports          2348
Misc            1750
Role-Playing    1500
Shooter         1323
Adventure       1303
Racing          1249
Platform         888
Simulation       874
Fighting         849
Strategy         683
Puzzle           580
Name: Genre, dtype: int64

In [44]:
#6) First five SNES games
df_games[df_games['Platform'] == 'SNES'].head(5)

Name Platform  Year_of_Release     Genre  \
18                      Super Mario World     SNES           1990.0  Platform   
56                  Super Mario All-Stars     SNES           1993.0  Platform   
71                    Donkey Kong Country     SNES           1994.0  Platform   
76                       Super Mario Kart     SNES           1992.0    Racing   
137  Street Fighter II: The World Warrior     SNES           1992.0  Fighting   

    Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  \
18   Nintendo     12.78      3.75      3.54         0.55         20.61   
56   Nintendo      5.99      2.15      2.12         0.29         10.55   
71   Nintendo      4.36      1.71      3.00         0.23          9.30   
76   Nintendo      3.54      1.24      3.81         0.18          8.76   
137    Capcom      2.47      0.83      2.87         0.12          6.30   

     Critic_Score  Critic_Count User_Score  User_Count Developer Rating  
18            NaN           NaN        NaN         NaN       NaN    NaN  
56            NaN           NaN        NaN         NaN       NaN    NaN  
71            NaN           NaN        NaN         NaN       NaN    NaN  
76            NaN           NaN        NaN         NaN       NaN    NaN  
137           NaN           NaN        NaN         NaN       NaN    NaN

In [45]:
#7) Five publishers with highest total global sales
df_games[['Publisher', 'Global_Sales']].groupby('Publisher').sum().sort_values(
    by = 'Global_Sales', ascending = False).head(5)

Global_Sales
Publisher                                
Nintendo                          1788.81
Electronic Arts                   1116.96
Activision                         731.16
Sony Computer Entertainment        606.48
Ubisoft                            471.61

In [46]:
#8) Create a new column that calculates the percentage of global sales from North America
df_games['NA_perc_sales'] = df_games['NA_Sales']/df_games['Global_Sales']
df_games.head(5)

Name Platform  Year_of_Release         Genre Publisher  \
0                Wii Sports      Wii           2006.0        Sports  Nintendo   
1         Super Mario Bros.      NES           1985.0      Platform  Nintendo   
2            Mario Kart Wii      Wii           2008.0        Racing  Nintendo   
3         Wii Sports Resort      Wii           2009.0        Sports  Nintendo   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  Critic_Score  \
0     41.36     28.96      3.77         8.45         82.53          76.0   
1     29.08      3.58      6.81         0.77         40.24           NaN   
2     15.68     12.76      3.79         3.29         35.52          82.0   
3     15.61     10.93      3.28         2.95         32.77          80.0   
4     11.27      8.89     10.22         1.00         31.37           NaN   

   Critic_Count User_Score  User_Count Developer Rating  NA_perc_sales  
0          51.0          8       322.0  Nintendo      E       0.501151  
1           NaN        NaN         NaN       NaN    NaN       0.722664  
2          73.0        8.3       709.0  Nintendo      E       0.441441  
3          73.0          8       192.0  Nintendo      E       0.476350  
4           NaN        NaN         NaN       NaN    NaN       0.359260

In [47]:
#9) Find the number of missing (NaN) entries in each column
df_games.isnull().sum()

Name                  2
Platform              0
Year_of_Release     269
Genre                 2
Publisher            54
NA_Sales              0
EU_Sales              0
JP_Sales              0
Other_Sales           0
Global_Sales          0
Critic_Score       8582
Critic_Count       8582
User_Score         6704
User_Count         9129
Developer          6623
Rating             6769
NA_perc_sales         0
dtype: int64

In [24]:
#9) Remove all rows with missing (NaN) entries
df_games_new = df_games.dropna()
df_games_new.shape

(6825, 17)

In [49]:
#10) Calculate the median user score and replace NaNs with the median score
#df_games['User_Score'].median()
#This give a cannot convert a string to float error
#Let's determine the string
df_games['User_Score'].unique()

array(['8', nan, '8.3', '8.5', '6.6', '8.4', '8.6', '7.7', '6.3', '7.4',
       '8.2', '9', '7.9', '8.1', '8.7', '7.1', '3.4', '5.3', '4.8', '3.2',
       '8.9', '6.4', '7.8', '7.5', '2.6', '7.2', '9.2', '7', '7.3', '4.3',
       '7.6', '5.7', '5', '9.1', '6.5', 'tbd', '8.8', '6.9', '9.4', '6.8',
       '6.1', '6.7', '5.4', '4', '4.9', '4.5', '9.3', '6.2', '4.2', '6',
       '3.7', '4.1', '5.8', '5.6', '5.5', '4.4', '4.6', '5.9', '3.9',
       '3.1', '2.9', '5.2', '3.3', '4.7', '5.1', '3.5', '2.5', '1.9', '3',
       '2.7', '2.2', '2', '9.5', '2.1', '3.6', '2.8', '1.8', '3.8', '0',
       '1.6', '9.6', '2.4', '1.7', '1.1', '0.3', '1.5', '0.7', '1.2',
       '2.3', '0.5', '1.3', '0.2', '0.6', '1.4', '0.9', '1', '9.7'],
      dtype=object)

From the output above, we see that there is a string 'tbd' in the critic score entries. We will replace this with NaN and then recalulate the median.

In [59]:
#10) continued
#replace 'tbd' with NaN
df_games['User_Score'] = df_games['User_Score'].replace('tbd', np.NaN)
#find the median user score
df_games['User_Score'].median()

7.5

In [70]:
#10) continued
#replace user score NaNs with median value
df_games['User_Score'].fillna(df_games['User_Score'].median(), inplace = True)
#display the new data frame
df_games.head(5)

Name Platform  Year_of_Release         Genre Publisher  \
0                Wii Sports      Wii           2006.0        Sports  Nintendo   
1         Super Mario Bros.      NES           1985.0      Platform  Nintendo   
2            Mario Kart Wii      Wii           2008.0        Racing  Nintendo   
3         Wii Sports Resort      Wii           2009.0        Sports  Nintendo   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  Critic_Score  \
0     41.36     28.96      3.77         8.45         82.53          76.0   
1     29.08      3.58      6.81         0.77         40.24           NaN   
2     15.68     12.76      3.79         3.29         35.52          82.0   
3     15.61     10.93      3.28         2.95         32.77          80.0   
4     11.27      8.89     10.22         1.00         31.37           NaN   

   Critic_Count User_Score  User_Count Developer Rating  NA_perc_sales  
0          51.0          8       322.0  Nintendo      E       0.501151  
1           NaN        7.5         NaN       NaN    NaN       0.722664  
2          73.0        8.3       709.0  Nintendo      E       0.441441  
3          73.0          8       192.0  Nintendo      E       0.476350  
4           NaN        7.5         NaN       NaN    NaN       0.359260